### PR TITLE
Add support for naked functions

### PR DESCRIPTION
**Pull request by ticki**
See https://github.com/rust-lang/rfcs/pull/1201#issuecomment-199442239

This PR adds `#[naked]` for marking naked functions.

### CODE REVIEW

**Comment by Ticki** [comment]
The tests are stolen from @jackpot51's PR.

**Comment by durka** [comment]
tidy failure

```
check: formatting
/home/travis/build/rust-lang/rust/src/test/codegen/naked-functions.rs:61: line longer than 100 chars
/home/travis/build/rust-lang/rust/src/libsyntax/feature_gate.rs:215: tab character
```

**Comment by Ticki** [comment]
Fixed.

File: src/libsyntax/feature_gate.rs [hunk 1]
```diff
@@ -376,6 +379,9 @@ pub const KNOWN_ATTRIBUTES: &'static [(&'static str, AttributeType, AttributeGat
     // FIXME: #14406 these are processed in trans, which happens after the
     // lint pass
     ("cold", Whitelisted, Ungated),
+    ("naked", Whitelisted, Gated("naked_functions",
+                                 "the `#[naked]` attribute \
+                                  is an experimental feature")),
```
Comment (eddyb) [inline src/libsyntax/feature_gate.rs]: Can you add a test that the feature gating works, i.e. you get an error by default?

Comment (Ticki) [inline src/libsyntax/feature_gate.rs]: This is fixed now.

File: src/test/compile-fail/gated-naked_functions.rs [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#[naked]
+fn naked() {
+    //~^ the `#[naked]` attribute is an experimental feature
```
Comment (eddyb) [inline src/test/compile-fail/gated-naked_functions.rs]: This doesn't need `ERROR`?

Comment (Ticki) [inline src/test/compile-fail/gated-naked_functions.rs]: [Apperently not](https://github.com/Ticki/rust/blob/master/src/test/compile-fail/gated-associated_consts.rs#L20).

Comment (eddyb) [inline src/test/compile-fail/gated-naked_functions.rs]: Fair enough.

Comment (eddyb) [inline src/test/compile-fail/gated-naked_functions.rs]: Fair enough. Travis failure suggests it needs to be placed on the attribute, though.

File: src/libsyntax/feature_gate.rs [hunk 3]
```diff
@@ -212,6 +212,9 @@ const KNOWN_FEATURES: &'static [(&'static str, &'static str, Option<u32>, Status
     // rust runtime internal
     ("unwind_attributes", "1.4.0", None, Active),
 
+    // allow the use of `#[naked]` on functions.
+    ("naked_functions", "1.9.0", None, Active),
```
Comment (durka) [inline src/libsyntax/feature_gate.rs]: I think this should be `Some(32408)`

**Comment by Ticki** [comment]
When will this be merged?

**Comment by TimNN** [comment]
@Ticki You can see the queue of PR to be tested & merged here: http://buildbot.rust-lang.org/homu/queue/rust

**Comment by eddyb** [comment]
@Ticki you might want to fix the nit @durka mentioned, but other than that it's waiting in the queue that @TimNN linked.

**Pull request merged**